File: solutions/ruby/pig-latin/2/pig_latin.rb
module PigLatin
  def self.translate(sentence)
    sentence.split.map{|word| translate_word(word)}.join(' ')
  end

  def self.translate_word(word)
    word.scan(/^(xr|yt|[aeiou])/) {return word + 'ay'}
    word.scan(/^(qu[aeiou])/) {return word[2..] + 'quay'}
    word.scan(/^([^aeio]u)/) {return word[1..] + 'ray'}
    word.scan(/^([^aeiou]+|.)y(.*)/) {return "y#{$2}#{$1}ay"}
    word.scan(/^(th[^aeiou]+)(.*)/) {return "#{$2}#{$1}ay"}
    word.scan(/^([^aeio]+)(.*)/) {return "#{$2}#{$1}ay"}
  end
end

File: solutions/python/pig-latin/2/pig_latin.py
import re


def translate_word(word: str) -> str:
    if re.match(r'^(?:yt|xr)|[aeiou]', word) is not None:
        return word + 'ay'

    re_match = re.match(r'^(qu+)(.*)', word)
    if re_match:
        return re_match.group(2) + re_match.group(1) + 'ay'

    re_match = re.match(r'^([^aeiouy]+)(.*)', word)
    if re_match:
        if re_match.group(2).startswith('ua'):
            return re_match.group(2)[1:] + re_match.group(1) + 'uay'
        return re_match.group(2) + re_match.group(1) + 'ay'

    re_match = re.match(r'^([^aeiou]+)(.*)', word)
    if re_match:
        return re_match.group(2) + re_match.group(1) + 'ay'

    return word


def translate(text: str) -> str:
    return ' '.join(map(translate_word, text.split()))
